webrtc/README.md

```diff
@@ -0,0 +1,323 @@
+
+- Can two _Browsers_ exchange their SDP packets via a third server node using
+  Circuit Relay v2 and DCUtR? Instead of exchanging the original SDP packets,
+  could they exchange their multiaddr and construct the remote's SDP packet
+  based on it?
+
+## Connection Security
+
+While WebRTC offers confidentiality and integrity via TLS, one still needs to
+authenticate the remote peer by its libp2p identity.
+
+After [Connection Establishment](#connection-establishment):
+
+1. _A_ opens a WebRTC datachannel and starts a Noise handshake using _A_'s and
+   _B_'s libp2p identity. See
+   [noise-libp2p](https://github.com/libp2p/specs/tree/master/noise).
+
+2. _A_ and _B_ write their TLS certificate fingerprint on the negotiated Noise
+   channel as multibase encoded multihashes (see [#addressing]).
+
+3. _A_ and _B_ read the other sides TLS certificate fingerprint on the
+   negotiated Noise channel and compare it to the ones verified during the DTLS
+   handshake.
+
+   Note: WebRTC supports different hash functions to hash the TLS certificate
+   (see https://datatracker.ietf.org/doc/html/rfc8122#section-5). The hash
+   function used in WebRTC, the hash function used for the hash exchanged in the
+   additional Noise handshake and the hash function used in the multiaddr
+   `/certhash` component MUST be the same. On mismatch the final Noise handshake
+   MUST fail.
+
+### Open Questions
+
+- Can a _Browser_ access the fingerprint of its TLS certificate?
+
+  Chrome allows you to access the fingerprint of any locally-created certificate
+  directly via `RTCCertificate#getFingerprints`. Firefox does not allow you to
+  do so. Browser compatibility can be found
+  [here](https://developer.mozilla.org/en-US/docs/Web/API/RTCCertificate). In
+  practice, this is not an issue since the fingerprint is embedded in the local
+  SDP string.
+
+- Is the above proposed additional handshake secure? See also alternative
+  proposed Handshake for
+  [WebTransport](https://github.com/libp2p/specs/pull/404).
+
+- Would it be more efficient for _B_ to initiate the Noise handshake? In other
+  words, who is able to write on an established WebRTC connection first? _A_ or
+  _B_?
+
+- Instead of exchanging the TLS fingerprints once the Noise handshake finished,
+  could one instead attach the fingerprints to the Noise handshake messages as
+  additional payloads? That would reduce the overall connection establishment
+  latency by one round trip. Would this schema be secure using the Noise XX
+  handshake pattern?
+
+- On the server side, can one derive the TLS certificate in a deterministic way
+  based on a node's libp2p private key? Benefit would be that a node only needs
+  to persist the libp2p private key and not the TLS key material while still
+  maintaining a fixed TLS certificate fingerprint.
+
+## Multiplexing
+
+After [Connection Security](#connection-security):
+
+1. On success of the authentication handshake _X_, the used datachannel is
```

Comment:
@mxinden WebRTC datachannel does not support closing reads and writes independently. The current draft implementations open new datachannels for each stream. While we should avoid a situation where we use noise over DTLS, we should consider using one data channel for auth, followed by a separate datachannel for transfer of data. This would allow us to use the existing multiplexers and upgraders over the second datachannel and would ease implementation. I'm specifically considering the ease of implementation on for the JavaScript case.

Reply:
> WebRTC datachannel does not support closing reads and writes independently.

Is the behavior documented in detail somewhere? After calling `RTCDataChannel.close` can one still read from the channel?

As an additional datapoint, rust-libp2p only exposes the ability to close the writing side of a stream, not the reading side. See [`Substream`](https://docs.rs/libp2p/latest/libp2p/core/trait.StreamMuxer.html#associatedtype.Substream).

If I interpret [`MuxedStream`](https://pkg.go.dev/github.com/libp2p/go-libp2p-core/network#MuxedStream) correctly, go-libp2p allows both closing the reading and closing the writing end today.

For the record, the corresponding excerpt from the current libp2p specification:

> They can also be "half closed", meaning that a stream can be closed for writing data but still open to receiving data and vice versa.

https://github.com/libp2p/specs/tree/master/connections#definitions

> followed by a separate datachannel for transfer of data. This would allow us to use the existing multiplexers and upgraders over the second datachannel and would ease implementation.

Using an additional multiplexer on top of a single WebRTC datachannel would be very unfortunate for the following reasons:

- We would have [head of line blocking](https://en.wikipedia.org/wiki/Head-of-line_blocking) given that the additional multiplexer runs on top of a single ordered WebRTC datachannel.
- The best general purpose muxer we have today is Yamux. Unfortunately Yamux fails short in some properties, e.g. backpressure on stream creation. See also https://github.com/libp2p/specs/pull/394.

> I'm specifically considering the ease of implementation on for the JavaScript case.

Can you expand on this? What makes the JavaScript side special?

Reply:
We can read any buffered messages and any buffered writes will be sent before closing. We cannot half close the channel. Refer: https://w3c.github.io/webrtc-pc/#closing-procedure
The current go-libp2p implementation is incorrect, since it only changes local state when closing reads/writes. 

The javascript implementation uses separate muxers and encrypters. Using a single datachannel would ease implementation as we could simply implement the `MultiaddrConnection` interface and use the existing upgraders and nop encrypter to return a muxed connection. 

If we consider using a single datachannel for signalling half close events, we cannot guarantee that the message signaling closure of read/write will be received after all other messages on the target data channel have been received.

Reply:
Just a datapoint: webrtc-rs seems to support half-closing a channel, see [`Stream`](https://docs.rs/webrtc-sctp/0.5.0/webrtc_sctp/stream/struct.Stream.html#method.shutdown) docs. @melekes to correct me in case I am wrong. Though this would still leave us with the constraint on the browser side.

---

Summarizing the different options to proceed thus far:

- Loosen the general libp2p requirement of supporting half-close.
  - Large effort and intrusive
- Nested multiplexing, e.g. running Yamux on top of SCTP.
  - Inefficient double muxing
  - We are again stuck with Yamux
  - Head of line blocking
- Implementing half-close on top of a WebRTC Datachannel, e.g. by message framing (`CloseRead`, `CloseWrite`, `Data([u8])`).
  - Overhead due to framing
  - Additional (mental) complexity

Reply:
Looking into the source for the rust stream, it looks like the `shutdown` call makes changes to local state and no additional signaling takes place between the peers.

Reply:
> Looking into the source for the rust stream, it looks like the `shutdown` call makes changes to local state and no additional signaling takes place between the peers.

that's correct. signalling only occurs when both halves of the stream are closed

Reply:
No progress from my end thus far. Some additional pointers:

- Spec
  - Streams should support both closing the writing and the reading side.
  - https://github.com/libp2p/specs/blob/master/connections/README.md#definitions
- Yamux
  - Support for closing the writing side of a stream
  - https://github.com/hashicorp/yamux/blob/master/spec.md#closing-a-stream
- Mplex
  - Support for closing the writing side of a stream
  - https://github.com/libp2p/specs/blob/master/mplex/README.md#closing-a-stream
- QUIC
  - Support for closing both writing and reading side of a stream
  - https://www.rfc-editor.org/rfc/rfc9000.html#name-bidirectional-stream-states

Reply:
@mxinden What if we used the DataChannel's `onmessage` API, and used the first byte for signaling stream state. We could just add the flags to this spec.

Reply:
> @mxinden What if we used the DataChannel's `onmessage` API, and used the first byte for signaling stream state. We could just add the flags to this spec.

Yes. That would be one option.

---

I am still eying another solution, namely _loosen the general libp2p requirement of supporting half-closed streams_. I am having a hard time coming up with use-cases where we would actually need half-closed streams. E.g. instead of signaling that one is no longer sending data on a stream through a half-close of the write-side, why not encode this at the application protocol level by sending a _no-longer-sending-data_ application protocol message?  Without being aware of future use-cases, it is hard for me to justify the additional complexity of e.g. message framing.

Can folks think of use-cases that require the feature of half-closing streams? May those use-cases instead communicate this information on the application protocol level instead of requiring the half-closing-stream feature at the multiplexer level? @marten-seemann do you have any use-cases in mind?

Reply:
> I am still eying another solution, namely loosen the general libp2p requirement of supporting half-closed streams.

That outcome would make me very sad. I think of bidirectional streams as a combination of two unidirectional streams. libp2p currently doesn’t have unidirectional streams, but there’s been desire to add this feature at some point.
Naturally, if you think of a bidirectional streams as two unidirectional streams glued together, each side’s write side can be closed independently (the same goes for resets).

> why not encode this at the application protocol level by sending a no-longer-sending-data application protocol message?

That might work well on reliable transports, although you’d be re-implementing a commonly used feature in many application protocols. Really, this is a transport feature. On unreliable transports (QUIC, WebTransport) you want to be able to stop retransmissions, which means that you won’t get around involving the transport layer. It is sad that WebRTC, running on top of an unreliable transport, doesn’t make use of this opportunity (at least in the way browsers implement their APIs).

Reply:
> On unreliable transports (QUIC, WebTransport) you want to be able to stop retransmissions, which means that you won’t get around involving the transport layer.

Can you expand on this @marten-seemann? If one endpoint would close its write side and thus stop sending, why would there be the need for retransmission?

Reply:
Let’s be clear about terminology here: _Closing_ a stream means the normal termination: You’ve successfully finished transmitting all the data you wanted to transmit. _Resetting_ a stream means: You’ve started transmitting data on a stream, but you’ve changed your mind and want to abort transmission as soon as possible.

You’d stop retransmissions when _resetting_ the stream, not when _closing_.

Reply:
Got it. This is helpful. Thanks @marten-seemann for clarifying.

**Stream resets**

We can only implement stream resets via additional messages on top of a stream. Such messages would be delivered in order, i.e. any data sent before would be delivered before. Thus I don't see a way how we could design proper stream resets on top of the browsers WebRTC API.

With the above in mind, is there any benefit of adding a `RESET` message on top of a datachannel? Would closing the datachannel via the browser's API not do the job?

**Stream closing**

As described in https://github.com/libp2p/specs/pull/412#issuecomment-1209246438 the browser does support closing a datachannel via [`RTCDataChannel.close()`](https://developer.mozilla.org/en-US/docs/Web/API/RTCDataChannel/close). Though unfortunately that as well closes the reading side of the datachannel.

We could add a `CLOSE_WRITE` message on top of a datachannel, indicating that the local endpoint will no longer write to the datachannel, but may continue reading from the datachannel.

What would be a use-case where an application protocol would leverage this mechanism. In other words, in which scenario would an application protocol want to indicate to the remote that it is no longer writing to a channel, but may continue reading from the channel?

Reply:
**Stream resets**

> With the above in mind, is there any benefit of adding a RESET message on top of a datachannel? Would closing the datachannel via the browser's API not do the job?

I'd argue that distinguishing between normal and abnormal stream termination is absolutely crucial for probably any application protocol. Take the following example: I'm uploading a large file, and after a while, I change my mind and cancel the upload. The server needs to be able to detect this, otherwise it might process the partial file.
This is such a basic functionality that essentially every stream muxer (SPDY, QUIC, H3, WebTransport, Yamux, mplex, and even TCP itself, via RST) implement it. Not sure why WebRTC doesn't expose an API for this in the browser, I'd assume that data channels were never intended to be used as a general-purpose transport as we're trying to do it here?

>  Such messages would be delivered in order, i.e. any data sent before would be delivered before. Thus I don't see a way how we could design proper stream resets on top of the browsers WebRTC API.

That is sad, considering that the underlying protocol _should_ allow us to be smarter about retransmissions. But if the API doesn't allow it, then there's little we can do. Note that this situation is at least not worse than what we're living with in the TCP world.

**Stream closing**

> What would be a use-case where an application protocol would leverage this mechanism. In other words, in which scenario would an application protocol want to indicate to the remote that it is no longer writing to a channel, but may continue reading from the channel?

That's a very common scenario, and we don't have to look any further than HTTP: The client serializes it's request, sends the request headers and (optionally) the request body (might be a file it is uploading). It closes the stream to indicate that the entire file has been sent. Seeing the stream being closed is the signal for the server that the upload is complete. The server can then process the upload and send a (large?) response on the same stream.

Reply:
https://github.com/mxinden/specs/pull/1 proposes option (3), namely supporting half-close and reset of streams on top of WebRTC's datachannels. @ckousik @melekes @marten-seemann and everyone else: Let me know what you think.

Reply:
With https://github.com/mxinden/specs/pull/1 merged, I am resolving here. Thanks for the input everyone.